### PR TITLE
feat: refactor train utilities for dtensor policy v2

### PR DESCRIPTION
**Pull request by hemildesai**
https://app.devin.ai/review/NVIDIA-NeMo/RL/pull/1757 helpful for review

Nightlies:

dpo-llama3.1-8b-instruct-4n8g-fsdp2tp2-quick.v2: https://wandb.ai/nvidia/nemo-rl/runs/fjrjhcsc

grpo-moonlight-16b-automodel-1n8g-ep8: https://wandb.ai/nvidia/nemo-rl/runs/gq9wibah

grpo-qwen2.5-7b-instruct-4n8g-fsdp2tp4.v3: https://wandb.ai/nvidia/nemo-rl/runs/r8qaeedc

grpo-qwen3-8B-base-1n8g-fsdp2-lora: https://wandb.ai/nvidia/nemo-rl/runs/xf11uk3t

sft-gpt-oss-20b-1n8g-fsdp8ep8-automodel: https://wandb.ai/nvidia/nemo-rl/runs/qsoscd63

sft-llama3.1-8b-1n8g-fsdp2tp1-lora: https://wandb.ai/nvidia/nemo-rl/runs/n5anaa2b

### Issues
#1589 

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Release Notes

* **New Features**
  * Introduced centralized training utilities with modular post-processing framework supporting loss computation, log-probability extraction, top-k logits, and score calculation
  * Enhanced training infrastructure with improved support for multimodal inputs and advanced parallel configurations
  * Added validation for model architecture consistency

* **Tests**
  * Added comprehensive test coverage for training components and distributed execution scenarios

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: nemo_rl/models/automodel/train.py [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+                    placements=[Shard(sequence_dim)],
+                )
+                break
+
+    return seq_index_dtensor, mb
+
+
+def forward_with_post_processing_fn(
+    model: nn.Module,
+    cfg: dict[str, Any],
```
Comment (terrykong) [inline nemo_rl/models/automodel/train.py]: can we narrow type to the policy config?

File: nemo_rl/models/automodel/train.py [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+        return outputs
+    elif not hasattr(outputs, "logits"):
+        return model.lm_head(outputs.last_hidden_state)
+    else:
+        return outputs.logits
+
+
+def apply_temperature_scaling(
+    logits: torch.Tensor,
+    cfg: dict[str, Any],
```
Comment (terrykong) [inline nemo_rl/models/automodel/train.py]: same policy config question

File: nemo_rl/models/automodel/train.py [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+    Returns:
+        tuple: (result, metrics, processed_microbatch)
+            - result: Output from post-processing (loss, logprobs, topk, or scores)
+            - metrics: Dictionary of metrics from post-processing
+            - processed_microbatch: The ProcessedMicrobatch that was processed
+    """
+    # Get the pre-processed microbatch from iterator or use provided one
+    if processed_mb is None:
+        if data_iterator is None:
+            raise ValueError("Either data_iterator or processed_mb must be provided")
```
Comment (terrykong) [inline nemo_rl/models/automodel/train.py]: should you do an xor check here to avoid silently ignoring?

File: nemo_rl/models/automodel/train.py [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+                break
+
+    return seq_index_dtensor, mb
+
+
+def forward_with_post_processing_fn(
+    model: nn.Module,
+    cfg: dict[str, Any],
+    post_processing_fn: PostProcessingFunction,
+    data_iterator: Optional[Iterator[ProcessedMicrobatch]] = None,
```
Comment (terrykong) [inline nemo_rl/models/automodel/train.py]: actually is data_iterator ever used? it seems like we'll always call this with processed_mb since we'll loop the iterator somewhere else. if not used, can we remove?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 5]
```diff
@@ -348,173 +375,41 @@ def train(
...
-
-                    if num_valid_samples > 0:
-                        mb_losses.append(loss.item())
-                        all_mb_metrics.append(loss_metrics)
+                # Extract losses and metrics from results
+                mb_losses = []
+                for mb_idx, (loss, loss_metrics) in enumerate(mb_results):
+                    # Only process valid (non-dummy) batches for metrics
+                    if mb_idx < iterator_len:
+                        num_valid_samples = loss_metrics.get("num_valid_samples", 0)
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: can this just be
```suggestion
                        num_valid_samples = loss_metrics["num_valid_samples"]
```
?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 6]
```diff
@@ -317,6 +309,41 @@ def train(
...
+                f"Emptying cache every {empty_cache_steps} microbatches, doing so unnnecessarily would incur a large performance overhead."
+            )
+
+        def on_microbatch_start(mb_idx):
+            if empty_cache_steps and mb_idx % empty_cache_steps == 0:
+                torch.cuda.empty_cache()
+
         with ctx:
             # Get data from batch and move to device
             data.to("cuda")
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: nit: i know it's equivalent for the batcheddatadict, but can we also just assign for consistency
```suggestion
            data = data.to("cuda")
```

**Review by terrykong — COMMENTED** [review]
thanks for all your efforts refactoring:

1. does this PR replace https://github.com/NVIDIA-NeMo/RL/pull/1815 ?
2. can we narrow any dict that is a policyconfig?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 7]
```diff
@@ -317,6 +309,41 @@ def train(
...
+                cp_buffers=processed_inputs.cp_buffers,
+                sequence_dim=sequence_dim,
+                dtype=self.dtype,
+                autocast_enabled=self.autocast_enabled,
+            )
+
+        # Setup cache clearing callback if configured
+        empty_cache_steps = self.cfg.get("dtensor_cfg", {}).get(
+            "clear_cache_every_n_steps"
+        )
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:385]: I saw a similar param `empty_unused_memory_level` in mcore, shall we using a same param and keep a same behavior?

also if we unify to a same param, shall we put it just under policy in config like `policy.xxx` instead of put it under `dtensor_cfg` or `megatron_cfg`?

cc @ananthsub

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 8]
```diff
@@ -348,173 +375,41 @@ def train(
...
+                    global_valid_toks=global_valid_toks,
+                    sequence_dim=sequence_dim,
+                    dp_size=self.dp_size,
+                    cp_size=self.cp_size,
+                    num_global_batches=num_global_batches,
+                    train_context_fn=train_context_fn,
+                    num_valid_microbatches=iterator_len,
+                    on_microbatch_start=on_microbatch_start
+                    if empty_cache_steps
+                    else None,
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: `empty_cache_steps` will be checked at `on_microbatch_start`, so feels no need to check here?

also remove this would be helpful if we want to do something else in `on_microbatch_start` later.
```suggestion
                    on_microbatch_start=on_microbatch_start
```

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: do you think it's better to move the statistic part to a function as well?

cc @ananthsub megatron as well

File: nemo_rl/models/automodel/train.py [hunk 9]
```diff
@@ -0,0 +1,869 @@
...
+
+    def __init__(
+        self,
+        loss_fn: LossFunction,
+        cfg: dict[str, Any],
+        device_mesh: Any,
+        cp_mesh: Any,
+        tp_mesh: Any,
+        cp_size: int,
+        dp_size: int,
```
Comment (yuki-97) [inline nemo_rl/models/automodel/train.py:459]: is it possible to keep a similar args with Megatron backend?

for these four classes
```
LogprobsPostProcessor,
LossPostProcessor,
ScorePostProcessor,
TopkLogitsPostProcessor,
```

https://github.com/NVIDIA-NeMo/RL/pull/1654/changes#diff-37e22992a045dc9e46d7194136eb20358b4078656bf9e1c86261735fcd93eb31R239-R244

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 10]
```diff
@@ -623,195 +521,32 @@ def get_logprobs(
...
-                input_lengths = lp_batch.get("input_lengths")
-                batch_size = input_ids.shape[0]
-
-                # Create post_attention_mask for right-padded data (used for masking after forward)
-                if not self.enable_seq_packing:
-                    post_attention_mask = torch.zeros(
-                        (batch_size, seq_len), dtype=torch.bool, device=input_ids.device
-                    )
-                    for i, length in enumerate(input_lengths):
-                        # For right-padded sequence, set 1s at the beginning of the sequence
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:706]: feels better to keep the comments?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 11]
```diff
@@ -960,161 +681,27 @@ def get_topk_logits(
...
-                            )
-                        else:
-                            full_logits = logits.to(torch.float32)
-                            vals, idx = torch.topk(full_logits, k=k, dim=-1)
-
-                # Handle sequence packing unpacking
-                if self.enable_seq_packing:
-                    # Unpack top-k results from packed format back to original batch format
-                    # vals: [1, packed_seq_len, k] -> [original_batch_size, original_seq_len, k]
-                    # idx: [1, packed_seq_len, k] -> [original_batch_size, original_seq_len, k]
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:1137]: feels better to keep the comments?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 12]
```diff
@@ -960,161 +681,27 @@ def get_topk_logits(
...
-                    # Get cumulative sequence lengths for unpacking
-                    cu_seqlens = flash_attn_kwargs.cu_seqlens_q
-
-                    for i in range(original_batch_size):
-                        start = cu_seqlens[i].item()
-                        end = cu_seqlens[i + 1].item()
-                        seq_len_actual = input_lengths[i].item()
-
-                        # Extract the corresponding portion from packed results
-                        # Note: vals and idx are [1, packed_seq_len, k] due to packing
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:1160]: feels better to keep the comments?

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py [hunk 13]
```diff
@@ -623,195 +521,32 @@ def get_logprobs(
...
-                                log_probs = torch.nn.functional.log_softmax(
-                                    logits, dim=-1
-                                )
-                            # Extract logprobs for each token in the sequence by gathering the logprob
-                            # corresponding to the next token at each position
-                            # Input shapes:
-                            #   log_probs: [batch_size, sequence_length, vocab_size] - logits for each position
-                            #   token_ids: [batch_size, sequence_length] - actual tokens
-                            # Output shape: [batch_size, sequence_length] - logprob of each token given previous
-                            # We get logprob of token[t+1] from logits[t], prepending 0 to maintain sequence length
```
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:826]: feels better to keep the comments?

**Review by yuki-97 — COMMENTED** [review]
thank you so much for the refactor efforts! just one last step to complete. 🎉

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py
Comment (yuki-97) [inline nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: is it solved in this PR? @RayenTian @hemildesai 

if so, we could remove the comment, and remove line 567 for `def score(...)` as well.
if not, is there anything plan to do? @RayenTian

Comment (RayenTian) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Yes, I think this PR can solve this. We can remove the comment and close the issue.

Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 10 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:706]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 11 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:1137]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 12 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:1160]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 13 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:826]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 7 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py:385]: will be solved in https://github.com/NVIDIA-NeMo/RL/issues/1888

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 6 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 5 above)
Comment (hemildesai) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: Done

File: nemo_rl/models/automodel/train.py (hunk 4 above)
Comment (hemildesai) [reply nemo_rl/models/automodel/train.py]: Removed data_iterator

File: nemo_rl/models/automodel/train.py (hunk 3 above)
Comment (hemildesai) [reply nemo_rl/models/automodel/train.py]: Updated.

File: nemo_rl/models/automodel/train.py (hunk 2 above)
Comment (hemildesai) [reply nemo_rl/models/automodel/train.py]: Done

File: nemo_rl/models/automodel/train.py (hunk 1 above)
Comment (hemildesai) [reply nemo_rl/models/automodel/train.py]: Done

File: nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py (hunk 8 above)
Comment (terrykong) [reply nemo_rl/models/policy/workers/dtensor_policy_worker_v2.py]: fixed in https://github.com/NVIDIA-NeMo/RL/pull/1757/changes/bca57d5816f6dc54618ca06e8cfbc1f5d49c815f

**Review by terrykong — APPROVED** [review]
